### PR TITLE
Push max threads to little less than kernel limit

### LINKED ISSUE #4990: fatal error: thread exhaustion
## Expected Behavior

nothing

## Current Behavior

server down，print "runtime: program exceeds 10000-thread limit fatal error: thread exhaustion"

## Steps to Reproduce (for bugs)

1. start 4 Java process ,each process init a singleton MinioClient
2. create 32 threads each process, to upload images to minio server
3. upload speed is about 10,000 pic/min 
4. api used was minioClient.putObject(bucketName,key,localFilePath);
5. after about 3 hours ,minio server went down,and loged "fatal error: thread exhaustion"

## Your Environment

* Version used (`minio version`):RELEASE.2017-08-05T00-00-53Z
* java sdk version : 3.0.7
* Operating System and version :centOS 7
* client and server are not on the same machine

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
Let Minio server use more threads than allowed by golang runtime. This
is important to better deal with high load.

## Motivation and Context
Fixes #4990 

## How Has This Been Tested?
Manually, by printing log

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/server-rlimit.go [hunk 1]
```diff
@@ -16,9 +16,33 @@
...
 
 func setMaxResources() (err error) {
+	// Set the Go runtime max threads threshold to 90% of kernel setting.
+	// Do not return any error when encountered since it is not a crucial task.
+	if runtime.GOOS == "linux" {
+		sysMaxThreadsStr, rErr := ioutil.ReadFile("/proc/sys/kernel/threads-max")
+		if rErr == nil {
+			sysMaxThreads, rErr := strconv.Atoi(strings.TrimSpace(string(sysMaxThreadsStr)))
+			if rErr == nil {
+				minioMaxThreads := (sysMaxThreads / 100) * 90
```
Comment (fwessels) [inline cmd/server-rlimit.go]: Minor comment: I would multiply first and then divide (less changes of some errors) and the size of the int is large enough

**Review by fwessels — APPROVED** [review]
LGTM

File: cmd/server-rlimit.go [hunk 2]
```diff
@@ -16,9 +16,33 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/minio/minio/pkg/sys"
+)
 
 func setMaxResources() (err error) {
+	// Set the Go runtime max threads threshold to 90% of kernel setting.
+	// Do not return any error when encountered since it is not a crucial task.
+	if runtime.GOOS == "linux" {
```
Comment (harshavardhana) [inline cmd/server-rlimit.go]: Lets move this to `pkg/sys` ? to have a platform dependent code in its own selective files? for example we can also use cgroup limits here as well.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by fwessels** [comment]
Looks like the issue referenced is incorrect (4490)

**Comment by vadmeste** [comment]
Can you review again @fwessels and @harshavardhana ?

**Review by harshavardhana — APPROVED** [review]

**Comment by deekoder** [comment]
@vadmeste  is looking into codecov marking blocked until then.

**Comment by vadmeste** [comment]
@deekoder, as @harshavardhana stated, codecov calculates the code coverage based on Windows & Linux code coverage.

Now, this PR adds a feature that only works for Linux, but not for Windows, that's why we can't enhance Windows code coverage here.

I think we can merge this since the code coverage for Linux is perfect.

**Pull request merged**